core/src/replay_stage.rs

```diff
@@ -645,58 +640,52 @@ impl ReplayStage {
     }
 
     #[allow(clippy::type_complexity)]
-    fn generate_votable_banks(
+    fn select_fork(
         ancestors: &HashMap<u64, HashSet<u64>>,
         bank_forks: &Arc<RwLock<BankForks>>,
         tower: &Tower,
         progress: &mut HashMap<u64, ForkProgress>,
-    ) -> Vec<(u128, Arc<Bank>, HashMap<u64, StakeLockout>, u64)> {
+    ) -> Option<(Arc<Bank>, u64)> {
         let tower_start = Instant::now();
         let frozen_banks = bank_forks.read().unwrap().frozen_banks();
+        let children = bank_forks.read().unwrap().descendants();
+
         trace!("frozen_banks {}", frozen_banks.len());
         let mut votable: Vec<(u128, Arc<Bank>, HashMap<u64, StakeLockout>, u64)> = frozen_banks
             .values()
-            .filter(|b| {
-                let has_voted = tower.has_voted(b.slot());
-                trace!("bank has_voted: {} {}", b.slot(), has_voted);
+            .filter(|bank| {
+                let has_voted = tower.has_voted(bank.slot());
+                trace!("bank has_voted: {} {}", bank.slot(), has_voted);
                 !has_voted
             })
-            .filter(|b| {
-                let is_locked_out = tower.is_locked_out(b.slot(), &ancestors);
-                trace!("bank is is_locked_out: {} {}", b.slot(), is_locked_out);
-                !is_locked_out
+            .filter(|bank| {
+                let leaf = children.get(&bank.slot()).map(|set| set.is_empty());
```

Comment:
we only ever want to vote on leaves?

Reply:
@rob-solana I think you are right.  If we vote on older heavier banks, we can vote on the child on the next round.  So this is redundant.